kdc/pkinit.c

```diff
@@ -458,6 +459,20 @@ _kdc_pk_rd_padata(astgs_request_t priv,
     hx509_verify_set_time(cp->verify_ctx, kdc_time);
     hx509_verify_attach_anchors(cp->verify_ctx, trust_anchors);
     hx509_certs_free(&trust_anchors);
+    
+    if (config->pkinit_kdc_revoke != NULL) {
+	hx509_revoke_init(context->hx509ctx, &revoke_ctx);
+	
+	ret = hx509_revoke_add_crl(context->hx509ctx, revoke_ctx, *config->pkinit_kdc_revoke);
```

Comment:
We already do this in `_krb5_pk_load_id()`, which gets called in `krb5_kdc_pk_initialize()`, which gets called in `krb5_kdc_pkinit_config()`, which gets called in `configure()`, which gets called in `main()` in the KDC.

Or was that not working?  And if, do you know why?

Reply:
Samba doesn't use the configure() (we start further in), but we do call krb5_kdc_pkinit_config(), so I'm not sure.

Reply:
Can you double-check?  Although this is harmless, since `hx509_revoke_add_crl()` will see that the CRL has already been loaded and return w/o doing anything.  If Samba is already calling `krb5_kdc_pkinit_config()` then this bit is not needed at all and I'd rather you drop it.

Reply:
Samba indeed calls `krb5_kdc_pkinit_config()`. The missing piece seems to be to use `hx509_verify_attach_revoke(cp->verify_ctx, kdc_identity->revokectx);` so that the revocation context from `_krb5_pk_load_id()` is used here. The only downside is that it would require a restart of Samba to reload the CRL file.

@abartlet Would it be possibile to let Samba redo the `krb5_kdc_pkinit_config()` on reload (for example through `smbcontrol smbd reload-config`?

Reply:
The revocation context does get attached in `_krb5_pk_load_id()`.

We should make it possible to reconfigure the KDC w/o restarting it, indeed.  I've been meaning to do something like that, and I'd like even to be able to store configurations in the HDB (and have done some of the work for that) so that we can reconfigure automatically via iprop (though, that does sound a bit scary).

Reply:
> The revocation context does get attached in `_krb5_pk_load_id()`.
It indeed does but without adding `hx509_verify_attach_revoke` to `_kdc_pk_rd_padata` I haven't been able to get revocation checking to work.

Reply:
`krb5_kdc_pkinit_config()` indeed calls `_krb5_pk_load_id()`, which sets up `verify_ctx` and `revokectx` on the static variable `kdc_identity`. However, as far as I can see, those contexts are never referenced again anywhere else. The verification context used in `_kdc_pk_rd_padata()` is an entirely different one that otherwise would have no revocation context attached.

So we ought either to call `hx509_verify_attach_revoke(cp->verify_ctx, kdc_identity->revokectx)`, as @wokis suggested in a previous comment, or to create a new revocation context as this MR does. But if we go with the latter option, we should add all the CRLs specified in the `pkinit_kdc_revoke` option, not just the first.

IMO the former option is preferable as it means less duplicated code. Reloading CRLs isn’t a problem, for there is already code to handle that: https://github.com/heimdal/heimdal/blob/f62e2f278437ff6c03d2d09bd628381c795bba78/lib/hx509/revoke.c#L812-L823

Reply:
Thus my suggested change would be: https://github.com/heimdal/heimdal/commit/aab2f8e5cb537d5a8241fba7bf9b7eb4c2e45fae